### ISSUE TITLE
gcp.compute.ForwardingRule breaks on latest pulumi/gcp 7.38.0   & pulumi/gcp  8.3.1

### ISSUE THREAD

**Issue by cl3nn** [pulumi/pulumi-gcp#2471]
### Describe what happened

When we update the Pulumi GCP provider from version 6 to either version 7 (7.38.0) or version 8 (8.3.1), we encounter an unusual error with the ForwardingRule resource during the pulumi preview step:

``` gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:509: sdk.helper_schema: missing expected {: provider=google-beta@8.3.1
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

Everything works correctly when using Pulumi GCP version 6.67.1."

### Sample program

problematic resource inside a ComponentResource:

```      
        this.forwardingRule = new gcp.compute.ForwardingRule(
            `${name}-fw-rule`,
            {
                target: this.regionHttpProxy.id,
                portRange: "80",
                region: args.region,
                loadBalancingScheme: "INTERNAL_MANAGED",
                network: args.networkId,
                subnetwork: this.subNetwork.id,
                allowGlobalAccess: true,
            },
            { parent: this },
        );
```
### Log output

when using pulumi/gcp 8.3.1

``` gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:509: sdk.helper_schema: missing expected {: provider=google-beta@8.3.1
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

when using pulumi/gcp 7.38.0

```gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:385: sdk.helper_schema: missing expected {: provider=google-beta@7.38.0
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

### Output of `pulumi about`

when using pulumi/gcp  8.3.1

```CLI          
Version      3.134.1
Go Version   go1.23.1
Go Compiler  gc

Plugins
KIND      NAME    VERSION
language  nodejs  unknown

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in nodejs: executable='/home/clenn/.nvm/versions/node/v20.11.1/bin/node' version='v20.11.1'
```

when using  pulumi/gcp  7.38.0

```CLI          
Version      3.134.1
Go Version   go1.23.1
Go Compiler  gc

Plugins
KIND      NAME    VERSION
resource  gcp     7.38.0
language  nodejs  unknown

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in nodejs: executable='/home/clenn/.nvm/versions/node/v20.11.1/bin/node' version='v20.11.1'`
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by cl3nn** [comment]
Hi @guineveresaenger  i cannot give you the complete stack state due to security reasons, but i've a redacted state of the working resource before the upgrade here:  https://gist.github.com/cl3nn/7d05759a43250567beb4a0f2c5515641

A minimal resource was given,  based on this example:  https://www.pulumi.com/registry/packages/gcp/api-docs/compute/forwardingrule/#forwarding-rule-http-lb

**Comment by guineveresaenger** [comment]
hey @cl3nn - okay I think I have a workaround for you. 🤞 

I managed to reproduce it on a [smallerloadbalancer example](https://www.pulumi.com/registry/packages/gcp/api-docs/compute/forwardingrule/#forwarding-rule-externallb); might be handy for testing.

Basically, what's happening is that v6 --> v7, the `serviceDirectoryRegistrations` type changed from a list with a single item to an object. For me, upgrading from v6.67.1 (newest v6) --> v7.38.0 (the version you tried to upgrade to), I could reproduce the error.
When you recreate the stack from scratch using v7, the stack file reads `"serviceDirectoryRegistrations": null,` not `"serviceDirectoryRegistrations": [],`, like it does for v6.

This is all especially frustrating of course for a user who isn't even using this field!

We do have means to prevent this sort of thing. It appears as though at least at some point we (or probably the upstream provider) had a working state migration from v6 --> v7, which was lost by the time v7.38.0 was published. I discovered this because I initially attempted to upgrade v6.67.1 --> v7.0.0, which succeeds. The state for the offending field is updated to be the expected null object.

**In summary:**

You should be able to upgrade today, without any changes, if you do it in two steps, avoiding v7.38.0:

1. upgrade v6.x --> v7.0.0
2. then upgrade --> v8.x

We will also very likely fix this universally to make sure we catch all state upgrades for everyone.
